### DISCUSSION TITLE
Possible to share developer notes of how scrcpy works

### DISCUSSION THREAD

**Discussion by krishtoautomate** [NetrisTV/ws-scrcpy#98] [Q&A]
scrcpy forwarding video is port in h264 format is understood. didn't get how touches are sent --

Which method or api is sending touches.
is touches sent to same that connect adb?
Which request body the scrcpy jar accepts for touch or keyboard commands.

Thank you

**Comment by drauggres** [comment]
>Which method or api is sending touches.

https://github.com/NetrisTV/ws-scrcpy/issues/56

> is touches sent to same that connect adb?

By default ws-scrcpy connects to scrcpy-server directly, without adb (unless `proxy over adb` was selected). For connection we use Web Sockets. Websockets are bidirectional, we receive video stream (and also video settings, clipboard) and send events.

> Which request body the scrcpy jar accepts for touch or keyboard commands.

Same answer as in #56:

All classes for `ControlMessage`s are here: https://github.com/NetrisTV/ws-scrcpy/tree/master/src/app/controlMessage. 
For messages format look at the `toBuffer` method.

On server side you should look [here](https://github.com/NetrisTV/scrcpy/blob/feature/websocket-v1.17.x/server/src/main/java/com/genymobile/scrcpy/ControlMessage.java) and [here](https://github.com/NetrisTV/scrcpy/blob/feature/websocket-v1.17.x/server/src/main/java/com/genymobile/scrcpy/ControlMessageReader.java).

**Comment by krishtoautomate** [comment]
if i use proxy over adb.
how touches are video buffer are communicated between device and computer?

**Comment by drauggres** [comment]
1. browser connects to the node.js server
2. server forwards port from device to localhost (same as `adb forward tcp:<FREE_FORT> tcp:8886`)
3. server creates proxy connection between browser and forwarded port

https://github.com/NetrisTV/ws-scrcpy/blob/67cc51c90861cadd1f977f14bc3efe8b5dfa56ca/src/server/goog-device/mw/WebsocketProxyOverAdb.ts#L49-L54

The rest is the same as with a direct connection.
